charts/kong/templates/certificate-kong-admin.yaml

```diff
@@ -0,0 +1,25 @@
+{{- if and ( .Capabilities.APIVersions.Has "cert-manager.io/v1" ) .Values.certificates.enabled .Values.certificates.admin.enabled }}
+# Kong Admin API and Manager Certificate
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: {{ include "kong.fullname" . }}-admin-cert
+  namespace:  {{ template "kong.namespace" . }}
+  labels:
+    {{- include "kong.metaLabels" . | nindent 4 }}
+spec:
+  secretName: {{ include "kong.fullname" . }}-admin-cert
+  commonName: {{ .Values.certificates.admin.commonName }}
+  dnsNames:
+  - {{ .Values.certificates.admin.commonName }}
+  {{- range .Values.certificates.admin.dnsNames }}
+  - {{ . | quote }}
+  {{- end }}
+  renewBefore: 360h
+  duration: 2160h
+  isCA: false
+  issuerRef:
+    name: {{ template "kong.certificate.issuer" . }}
+    kind: Issuer
```

Comment:
We'll want to rework this to use the user-provided issuer and to support either Issuers or ClusterIssuers. To that end, we probably want to add an additional `.certificates.SERVICE.clusterIssuer` field and top-level `.certificates.issuer` and `.certificates.clusterIssuer` fields (assuming many users probably don't need to configure a different one per-Service).

The template can then generate the complete `issuerRef`, with logic like:
```
if .certificates.SERVICE.clusterIssuer:
issuerRef:
   name: .certificates.SERVICE.clusterIssuer
   kind: ClusterIssuer
else if .certificates.SERVICE.issuer:
<similar>
else if .certificates.clusterIssuer:
<similar>
else if .certificates.issuer:
<similar>
else
<use the self-signed clusterissuer from https://cert-manager.io/docs/configuration/selfsigned/ as a final fallback>
```
I.e. prefer the specific issuer if set, else the global. Arbitrarily prefer clusterissuers because you shouldn't set both, but we'll try to generate valid YAML even if we get conflicting instructions (Helm doesn't provide a great way to reject logically wrong configuration).

You can optionally reduce these into a single generic template that's used for each Certificate. Similar to the [Ingress generator](https://github.com/Kong/charts/blob/dedd27073980e866bd97e2187600f1a60f9af453/charts/kong/templates/_helpers.tpl#L62-L111), you can define a `kong.certmanagercertificate` and pass it a non-root context, and then just use `{{ .commonName }}` instead of the full paths. In an ideal world you'd just pass `.Values.certificates.admin` as the context, but in practice, since we want to use fullname and such, you'll probably need to [build a context object while you're still at the root](https://github.com/Kong/charts/blob/dedd27073980e866bd97e2187600f1a60f9af453/charts/kong/templates/service-kong-portal.yaml) (Helm supposedly provides a way to get to the root from a non-root context, but I've never been able to get it to work). I'd recommend condensing to a single template now, but it's easy enough to do after, so if you'd prefer not to just create an issue to do it later.

Reply:
Added support for user defined top level, and service specific `issuer` and `clusterIssuer` [422b493](https://github.com/Kong/charts/pull/592/commits/422b4939f7eee498e86fbc37790b37cd84b9abbf).

Opened #654 to track template consolidation.

Reply:
> ```
> else
> <use the self-signed clusterissuer from https://cert-manager.io/docs/configuration/selfsigned/ as a final fallback>
> ```

I'm unsure it makes sense to have an `{{ else }}` fallback in the template if we do not also have a template to generate the namespaced selfsigned issuer with the release.

I would like to add a selfsigned issuer template that can deploy with the release as a fallback to support this `else` behavior.

I acknowledge that this issuer fallback would float up to a meta chart and be removed from this core kong chart once a meta chart is released.

Reply:
Added https://github.com/usrbinkat/kong-charts-feat-cert-manager/pull/1 for consolidation. Manual smoke test log at [test.txt](https://github.com/Kong/charts/files/9285431/test.txt). After that and a merge of upstream main/fixing conflicts, this should be good to go.

We can probably take a doc/umbrella approach instead of a fallback and just remove the `else`:

- In the umbrella chart, it includes an issuer and sets `kong.certificates.clusterIssuer: <issuer name>` in the default values.yaml.
- If installing alone, the readme provides concise instructions. Two settings where we can provide explicit values are not an undue burden.

The error you hit if you install with no fallback and no configuration isn't perfect, but is at least in the right ballpark:

```
Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: error validating "": error validating data: ValidationError(Certificate.spec): missing required field "issuerRef" in io.cert-manager.v1.Certificate.spec
```